### ISSUE TITLE
Suggestion: execute .m files

### ISSUE THREAD

**Issue by Jeff-Wolf** [Gimly/vscode-matlab#19]
Thanks for developing this extension!

To my limited knowledge - I'm a newbie to VScode and this package and will be happy to be told better - .m files cannot be executed at this point. (There is a pertinent suggestion in the code runner extension labeled with 'help needed'.)

I think it would be valuable for VSCode to have this feature.

**Comment by Gimly** [comment]
Definitely something that I'll look into, thanks for the idea.

**Comment by Jeff-Wolf** [comment]
Great - thanks! I am not familiar with json but I use matlab a lot. Will be happy to support you.

**Comment by Gimly** [comment]
@Jeff-Wolf Do you have any idea if there is already an open source extension that does this for matlab? Or know of a way to execute and debug Matlab code outside of the Matlab editor itself?

**Comment by Jeff-Wolf** [comment]
The one I know and work with is the script package for Atom. It does not allow to debug the Matlab code though.

**Comment by Jeff-Wolf** [comment]
Octave is open source and almost completely compatible with Matlab. The latest distributions do come with a GUI including and include a built-in debugger. This could be another reference to look into.

**Comment by Gimly** [comment]
OK quickly checked the script package for atom, all it does is calling the Matlab executable and getting back the output to display in the editor. Would that be sufficiently interesting as a feature?

Octave is a complete reimplementation of Matlab, so they have their own debugger and everything, if I would use Octave for debugging I would force all users to install Octave, which I don't think is a very good idea.

**Comment by Jeff-Wolf** [comment]
It would definitely meet my requirements. I still use the native Matlab IDE for most of my code development but like to run and make minor adjustments in environments which can also execute tex, R, etc. and have a smooth git integration.

I agree with your point regarding Octave. I installed it at some point but I doubt most Matlab users have it.

**Comment by ragavsathish** [comment]
Hi @Gimly , Thanks for this amazing plugin and @Jeff-Wolf,  I mostly use MATLAB IDE in windows environment but I really liked the way it works under Linux. In Linux, one can open the Matlab terminal in the VS code IDE and debug /run '.m' file in the terminal itself. 

On the other hand, Matclipse plugin is one of the examples that enables Matlab integration with IDE.  This eclipse plugin calls a remote instance of Matlab and queries the workspace variables and shows up the in the eclipse pannel.
https://github.com/krystophny/matclipse

**Comment by Jeff-Wolf** [comment]
Thanks for pointing this out @ragavsathish
I am not working with Matclipse and just saw that it requires to install several eclipse components. Same problem as for Octave might apply...

@Gimly Maybe it could start with a feature like script for Atom offers and then the debugger comes next? (I have no idea about how long it takes to implement these features.)

**Comment by Gimly** [comment]
Yeah, Matclipse looks quite interesting but it's a huge work. Maybe something to think in the long run, I'd be interested in finding resources directly from matlab about running an external debugger.

@ragavsathish I don't understand, you say that on Linux you can open the terminal in VS Code and run matlab directly from it, but what's blocking from doing the same on Windows?

**Comment by ragavsathish** [comment]
@Gimly At least in Windows 7, whenever I open Matlab in PowerShell or command prompt, it opens up a new command window (possibly JVM based ) like below. Hence I need to alternate between the windows rather having the command editor docked in VS CODE IDE which is a lot cooler and easier.
 
![image](https://user-images.githubusercontent.com/7013764/30983092-17884dd4-a492-11e7-9fca-638bbd457b25.png)

**Comment by Gimly** [comment]
@ragavsathish Could you try doing this:
```
matlab -nodisplay -nosplash -nodesktop -r "run('C:\your\path\mfile.m');"
```

**Comment by sco1** [comment]
@Gimly launching MATLAB from the command line is [by design fundamentally different](https://blogs.mathworks.com/community/2010/02/22/launching-matlab-without-the-desktop/) on [Windows](https://www.mathworks.com/help/matlab/ref/matlabwindows.html) and [Linux](https://www.mathworks.com/help/matlab/ref/matlablinux.html)/[Mac](https://www.mathworks.com/help/matlab/ref/matlabmac.html). Where the Linux/Mac versions support the `-nodesktop` option and will run natively in the terminal, the Windows version does not support the `-nodesktop` option, and will instead launch a command window rather than continue execution in the terminal. It's not necessarily feature-breaking, but it can certainly get annoying.

A related option for Windows could possibly be MATLAB's [COM Automation Server](https://www.mathworks.com/help/matlab/call-matlab-com-automation-server.html) or possibly MATLAB's [external engine APIs](https://www.mathworks.com/help/matlab/matlab-engine-applications.html), though I'm not familiar enough with either to determine whether or not they're useful in this situation.

**Comment by Gimly** [comment]
@sco1 Thanks a lot for those information, very useful! The fact that the Windows version doesn't support the `-nodesktop` option is pretty annoying. Maybe we can live with the java window for the first iteration though?

Going through the COM automation server could be interesting and allow for the debugging as well, but it's a Windows only option then, quite annoying for a multiplatform extension. Using the API is also an interesting option, but quite the challenge.

**Comment by Pexonazo** [comment]
Maybe just fork extension for Windows and Linux/Mac, since they're so different?

**Comment by urob** [comment]
One might be able to run Matlab interactively using the existing Jupyter extension:
https://marketplace.visualstudio.com/items?itemName=donjayamanne.jupyter

For a Matlab kernel for Jupyter see here:
https://github.com/imatlab/imatlab

**Comment by OsirisTerje** [comment]
Have there been any progress on this issue now ?

**Comment by Gimly** [comment]
@OsirisTerje Not really, it's pretty locked by the fact that we can't execute code through the command line on Windows without passing through the Java engine, so it cannot be ran correctly on VSCode's command line. Using the COM Automation would be a Windows only solution, which isn't really acceptable. The external engine API seems like a huge challenge...

In the end I'm not sure what would be the best approach for tackling this, maybe there is something we have overlooked?

**Comment by bersbersbers** [comment]
MATLAB offers a Python API - might this help?
https://de.mathworks.com/help/matlab/programming-interfaces-for-external-languages.html

**Comment by shikouzhong** [comment]
@Gimly Maybe it can be implemented through this way, though I'm not quite sure about it:
MATLAB can use the Java® debugger to do it. The Java debugger uses the default portnumber value 4444 to communicate with MATLAB (help from: https://ww2.mathworks.cn/help/matlab/ref/matlabwindows.html). And we know VS Code has good debug support for Java.

**Comment by rick3rt** [comment]
I have a (somewhat hacky) solution to run `.m` files in an open Matlab session on Windows. 

It involves an additional extension in VSCode and having to run AutoHotKey, but this way I am able to 

|Action|Shortcut|
|-|-|
|Pull the Command Window to the foreground by pressing | <kbd>Ctrl</kbd>+<kbd>m</kbd>|
|Run the currently open file by pressing | <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>Alt</kbd>+<kbd>m</kbd>|
|Run the current line by pressing (when no text is selected) | <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>m</kbd>|
|Run the current selection by pressing | <kbd>Ctrl</kbd>+<kbd>shift</kbd>+<kbd>m</kbd>|
|Run the current **section** by pressing | <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>m</kbd>|

If there is interest I can write a small guide on how to setup this (including the AutoHotKey script).

**Comment by GitDor** [comment]
@rick3rt I would love to have this in my VSCode. This would make my workflow much more efficient :)

**Comment by rick3rt** [comment]
@GitDor allright, I created a repo with the guide. Hope I didn't miss a step... https://github.com/rick3rt/vscode-run-matlab

Let me know what you think!

**Comment by Galileji** [comment]
> @GitDor allright, I created a repo with the guide. Hope I didn't miss a step... https://github.com/rick3rt/vscode-run-matlab
> 
> Let me know what you think!

This is amazing!
I slightly modified the auto hotkey mode so that matlab runs using F5 and F9 (for parts of code), which is what happens with Matlab editor. To preserve Vs code beahaviour with other kind of files, it is possible to allow these shortcuts only if the active tab contains ".m -".

```
SetTitleMatchMode, 2    ;2 means A window's title can contain WinTitle anywhere inside it to be a match

#ifWinActive, .m - ahk_exe Code.exe, ; window's title contain ".m -" and process is "Code.exe"
F5::ExecFileMatlab()          
F9::CopyExecMatlab()       
```
